tests/test_build_latex.py

```diff
@@ -814,3 +814,123 @@ def test_maxlistdepth_at_ten(app, status, warning):
+    assert ('\\hline\n'
+            '\\sphinxstylethead{\\relax \nheader1\n\\unskip}\\relax &'
+            '\\sphinxstylethead{\\relax \nheader2\n\\unskip}\\relax' in table)
+    assert ('\\hline\ncell1-1\n&\ncell1-2\n\\\\' in table)
+    assert ('\\hline\ncell2-1\n&\ncell2-2\n\\\\' in table)
+    assert ('\\hline\ncell3-1\n&\ncell3-2\n\\\\' in table)
+    assert ('\\hline\\end{tabulary}' in table)
+
+    # table having :widths: option
+    table = tables['table having :widths: option']
+    assert ('\\noindent\\begin{tabulary}{\\linewidth}{'
+            '|p{\\dimexpr(\\linewidth-\\arrayrulewidth)*30/100-2\\tabcolsep-\\arrayrulewidth\\relax}'
+            '|p{\\dimexpr(\\linewidth-\\arrayrulewidth)*70/100-2\\tabcolsep-\\arrayrulewidth\\relax}|}'
+            in table)
+    assert ('\\hline\n'
+            '\\sphinxstylethead{\\relax \nheader1\n\\unskip}\\relax &'
+            '\\sphinxstylethead{\\relax \nheader2\n\\unskip}\\relax' in table)
+    assert ('\\hline\ncell1-1\n&\ncell1-2\n\\\\' in table)
+    assert ('\\hline\ncell2-1\n&\ncell2-2\n\\\\' in table)
+    assert ('\\hline\ncell3-1\n&\ncell3-2\n\\\\' in table)
+    assert ('\\hline\\end{tabulary}' in table)
+
+    # table having caption
+    table = tables['table having caption']
+    assert ('\\begin{threeparttable}\n\\capstart\\caption{caption for table}'
+            '\\label{\\detokenize{index:id1}}' in table)
+    assert ('\\noindent\\begin{tabulary}{\\linewidth}{|L|L|}' in table)
+    assert ('\\hline\n'
+            '\\sphinxstylethead{\\relax \nheader1\n\\unskip}\\relax &'
+            '\\sphinxstylethead{\\relax \nheader2\n\\unskip}\\relax' in table)
+    assert ('\\hline\ncell1-1\n&\ncell1-2\n\\\\' in table)
+    assert ('\\hline\ncell2-1\n&\ncell2-2\n\\\\' in table)
+    assert ('\\hline\ncell3-1\n&\ncell3-2\n\\\\' in table)
+    assert ('\\hline\\end{tabulary}' in table)
+    assert ('\\end{threeparttable}' in table)
+
+    # table having verbatim
+    table = tables['table having verbatim']
+    assert ('\\noindent\\begin{tabular}{|*{2}{'
+            'p{\\dimexpr(\\linewidth-\\arrayrulewidth)/2-2\\tabcolsep-\\arrayrulewidth\\relax}|}}\n'
+            '\\hline' in table)
+
+    # table having problematic cell
+    table = tables['table having problematic cell']
+    assert ('\\noindent\\begin{tabular}{|*{2}{'
+            'p{\\dimexpr(\\linewidth-\\arrayrulewidth)/2-2\\tabcolsep-\\arrayrulewidth\\relax}|}}\n'
+            '\\hline' in table)
+
+    # table having both :widths: and problematic cell
+    table = tables['table having both :widths: and problematic cell']
+    assert ('\\noindent\\begin{tabular}{'
+            '|p{\\dimexpr(\\linewidth-\\arrayrulewidth)*30/100-2\\tabcolsep-\\arrayrulewidth\\relax}'
+            '|p{\\dimexpr(\\linewidth-\\arrayrulewidth)*70/100-2\\tabcolsep-\\arrayrulewidth\\relax}|}'
+            in table)
+
+    # longtable
+    table = tables['longtable']
+    assert ('\\begin{longtable}{|l|l|}\n\\hline' in table)
```

Comment:
@tk0miya  I am currently merging with latest update to master (#3381) and will push a branch to my fork based on yours so that you can fetch from there later to let tests pass.